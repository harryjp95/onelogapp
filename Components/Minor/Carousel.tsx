import fetch from 'node-fetch'
import React, { ReactElement, useEffect, useState } from 'react'
import { ActivityIndicator, ScrollView, StyleSheet, View } from 'react-native'

import { GradientBorder } from './GradientBorder'

interface CarouselProps {
  table: string,
  Template: ReactElement
}

const getReadings = (table: string) => {
  const url = `http://localhost:8088/readings/${table}`

  return fetch(url, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json'
    }
  })
    .then(res => res.json())
    .catch(err => err)
}

const Carousel: React.FC<CarouselProps> = (props: CarouselProps) => {
  const { table, Template } = props
  const [readings, setReadings] = useState()

  useEffect(() => {
    getReadings(table).then(res => setReadings(res.reverse()))
  }, [])

  let templates
  if (readings) {
    templates = readings.map((r, i) => {
      return <Template created={r.created} reading={r.reading} index={i} />
    })
  }

  return(
    readings
      ? <>
        <ScrollView horizontal={true} pagingEnabled={true}>
          { templates }
        </ScrollView>
        <GradientBorder x={1.0} y={1.0} />
        </>
      : <View style={{ flex: 1 }}>
          <View style={{ flex: 1 }} />
          <View style={{ ...Styles.lastReading }}>
            <ActivityIndicator color={'black'} />
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
  )
}

const Styles = StyleSheet.create({
  lastReading: {
    flex: 5,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center'
  }
})

export default Carousel
