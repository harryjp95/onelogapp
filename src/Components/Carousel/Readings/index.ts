export * as BgCarousel from './Bg'
export * as StatsCarousel from './Stats'
export * as DoseCarousel from './Dose'
export * as MacroCarousel from './Macro'
