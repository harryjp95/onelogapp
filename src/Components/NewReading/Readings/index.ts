export * from './Bg';
export * from './Dose';
export * from './Macro';
export * from './Keto';
