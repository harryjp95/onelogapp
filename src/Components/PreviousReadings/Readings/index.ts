export * from './PreviousBgReading';
export * from './PreviousKetoReading';
export * from './PreviousDoseReading';
export * from './PreviousMacroReading';
