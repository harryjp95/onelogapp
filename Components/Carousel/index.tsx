import React, { useState, ReactElement } from 'react'
import { ActivityIndicator, StyleSheet, Text, View } from 'react-native'
import GestureRecognizer from 'react-native-swipe-gestures';

import { GradientBorder } from '../Minor/GradientBorder'
import { Chevron } from '../Minor/Chevron'
import { generateCreatedDate } from '../Helpers/DateHelpers'
import { capitalise } from '../Helpers/GeneralHelpers'

interface CarouselProps {
  table: string,
  Template: React.FC,
  readings: ReactElement[]
}

const Carousel: React.FC<CarouselProps> = (props: CarouselProps) => {
  const { table, Template, readings } = props
  const [index, setIndex] = useState(0)

  const handleSwipeLeft = () => {
    if (index < readings.length - 1) setIndex(index + 1)
  }

  const handleSwipeRight = () => {
    if (index > 0) setIndex(index - 1)
  }

  const reading = readings && readings[index]

  return(
    readings.length > 0
      ? <View style={Styles.container}>
          <View style={Styles.header}>
            <Text style={Styles.tag}>
              { capitalise(table) }
            </Text>
            <Text style={Styles.time}>
              { generateCreatedDate(reading['created']) }
            </Text>
          </View>
          <GradientBorder x={0.4} y={1.0} colors={['grey', '#ebebeb']} />

          <View style={Styles.contentContainer}>
            <View style={Styles.chevron}>
            {index === 0
              ? <Chevron symbol={''} handlePress={() => null}/>
              : <Chevron symbol={'<'} handlePress={handleSwipeRight} />}
            </View>
            <GestureRecognizer onSwipeLeft={handleSwipeLeft} onSwipeRight={handleSwipeRight} style={Styles.template}>
              <Template data={reading} />
            </GestureRecognizer>
            <View style={Styles.chevron}>
            {index < readings.length - 1
              ? <Chevron symbol={'>'} handlePress={handleSwipeLeft} />
              : <Chevron symbol={''} handlePress={() => null} />}
            </View>
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
      : <View style={{ flex: 1 }}>
          <View style={{ flex: 1 }} />
          <View style={{ ...Styles.lastReading }}>
            <ActivityIndicator color={'black'} />
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
  )
}

const Styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#ebebeb'
  },
  header: {
    flex: 1,
    flexDirection: 'row',
    width: '100%',
  },
  contentContainer: {
    flexDirection: 'row',
    flex: 5,
    alignItems: 'center',
    justifyContent: 'center'
},
  template: {
    flex: 4,
    alignContent: 'center'
  },
  chevron: {
    flex: 1
  },
  tag: {
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'center',
    padding: 4,
    color: 'white',
    backgroundColor: 'black',
    width: 60
  },
  time: {
    fontSize: 16,
    fontWeight: 'bold',
    padding: 4
  },
  lastReading: {
    flex: 5,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center'
  }
})

export default Carousel
