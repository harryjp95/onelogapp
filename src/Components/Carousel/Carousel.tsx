import React, { useState } from 'react'
import { ActivityIndicator, Text, View } from 'react-native'
import GestureRecognizer from 'react-native-swipe-gestures'

import { Chevron, GradientBorder } from '../Minor'
import { capitalise, generateCreatedDate } from '../../Helpers'
import { Reading } from '../../types'
import { CarouselStyles } from './Styles'

interface ReadingCarouselProps {
  reading: Reading
}

interface CarouselProps {
  name: string
  Template: React.FC<ReadingCarouselProps>
  readings: any[]
  startingIndex?: number
}

const Carousel: React.FC<CarouselProps> = (props: CarouselProps) => {
  const { name, Template, readings, startingIndex } = props
  const [index, setIndex] = useState(startingIndex || 0)

  const handleSwipeLeft = () => {
    if (index < readings.length - 1) setIndex(index + 1)
  }

  const handleSwipeRight = () => {
    if (index > 0) setIndex(index - 1)
  }

  const reading = readings && readings[index]

  return readings && readings.length > 0 ? (
    <View style={CarouselStyles.container} testID="carousel">
      <View style={CarouselStyles.header}>
        <Text style={CarouselStyles.tag}>{capitalise(name)}</Text>
        <Text style={CarouselStyles.time}>{generateCreatedDate(reading.created) || ''}</Text>
      </View>
      <GradientBorder x={0.4} y={1.0} colors={['grey', '#ebebeb']} />

      <View style={CarouselStyles.contentContainer}>
        <View style={CarouselStyles.chevron}>
          {index === 0 ? <Chevron handlePress={() => null} /> : <Chevron left handlePress={handleSwipeRight} />}
        </View>
        <GestureRecognizer onSwipeLeft={handleSwipeLeft} onSwipeRight={handleSwipeRight} style={CarouselStyles.template}>
          <Template reading={reading} />
        </GestureRecognizer>
        <View style={CarouselStyles.chevron}>
          {index < readings.length - 1 ? (
            <Chevron right handlePress={handleSwipeLeft} />
          ) : (
            <Chevron handlePress={() => null} />
          )}
        </View>
      </View>
      <GradientBorder x={1.0} y={1.0} />
    </View>
  ) : (
    <View style={CarouselStyles.placeholder} testID="carousel">
      <View style={CarouselStyles.placeholder} />
      <View style={CarouselStyles.lastReading}>
        <ActivityIndicator color="black" />
      </View>
      <GradientBorder x={1.0} y={1.0} />
    </View>
  )
}

Carousel.defaultProps = {
  startingIndex: 0
}

export default Carousel
