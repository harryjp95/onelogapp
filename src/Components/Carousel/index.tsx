import React, { useEffect, useState } from 'react'
import { ActivityIndicator, StyleSheet, Text, View } from 'react-native'
import GestureRecognizer from 'react-native-swipe-gestures';

import { IBgReading } from './Readings/Bg'
import { IStatsReading } from './Readings/Stats'
import { IDoseReading } from './Readings/Dose'
import { IMacroReading } from './Readings/Macro';

import { getData } from '../Store/index'
import { GradientBorder } from '../Minor/GradientBorder'
import { Chevron } from '../Minor/Chevron'
import { generateCreatedDate } from '../Helpers/Date'
import { capitalise } from '../Helpers/General'

export interface BgTemplateProps {
  data: IBgReading
}
export interface StatsTemplateProps {
  data: IStatsReading
}
export interface DoseTemplateProps {
  data: IDoseReading
}
export interface MacroTemplateProps {
  data: IMacroReading
}

interface CarouselProps {
  table: string,
  Template: React.FC<BgTemplateProps> | React.FC<StatsTemplateProps> | React.FC<DoseTemplateProps> | React.FC<MacroTemplateProps>,
  dataKey: string
}

const Carousel: React.FC<CarouselProps> = (props: CarouselProps) => {
  const { table, Template, dataKey } = props
  const [readings, setReadings] = useState([] as any)
  const [index, setIndex] = useState(0)

  useEffect(() => {
    const fetchReadings = async (key: string) => {
      try {
        const data = await getData(key)
        if (data && data.readings) {
          setReadings(data.readings)
        }
      } catch(err) {
        console.log('Error fetchReadings: ', err)
      }
    }
    fetchReadings(dataKey)
  }, [])

  const handleSwipeLeft = () => {
    if (index < readings.length - 1) setIndex(index + 1)
  }

  const handleSwipeRight = () => {
    if (index > 0) setIndex(index - 1)
  }

  const reading = readings && readings[index]

  return(
    readings.length > 0
      ? <View style={Styles.container}>
          <View style={Styles.header}>
            <Text style={Styles.tag}>
              { capitalise(table) }
            </Text>
            <Text style={Styles.time}>
              { generateCreatedDate(reading['created']) }
            </Text>
          </View>
          <GradientBorder x={0.4} y={1.0} colors={['grey', '#ebebeb']} />

          <View style={Styles.contentContainer}>
            <View style={Styles.chevron}>
            {index === 0
              ? <Chevron symbol={''} handlePress={() => null}/>
              : <Chevron symbol={'<'} handlePress={handleSwipeRight} />}
            </View>
            <GestureRecognizer onSwipeLeft={handleSwipeLeft} onSwipeRight={handleSwipeRight} style={Styles.template}>
              <Template data={reading} />
            </GestureRecognizer>
            <View style={Styles.chevron}>
            {index < readings.length - 1
              ? <Chevron symbol={'>'} handlePress={handleSwipeLeft} />
              : <Chevron symbol={''} handlePress={() => null} />}
            </View>
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
      : <View style={{ flex: 1 }}>
          <View style={{ flex: 1 }} />
          <View style={{ ...Styles.lastReading }}>
            <ActivityIndicator color={'black'} />
          </View>
          <GradientBorder x={1.0} y={1.0} />
        </View>
  )
}

const Styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#ebebeb'
  },
  header: {
    flex: 1,
    flexDirection: 'row',
    width: '100%',
  },
  contentContainer: {
    flexDirection: 'row',
    flex: 5,
    alignItems: 'center',
    justifyContent: 'center'
},
  template: {
    flex: 4,
    alignContent: 'center'
  },
  chevron: {
    flex: 1
  },
  tag: {
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'center',
    padding: 4,
    color: 'white',
    backgroundColor: 'black',
    width: 60
  },
  time: {
    fontSize: 16,
    fontWeight: 'bold',
    padding: 4
  },
  lastReading: {
    flex: 5,
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'center'
  }
})

export default Carousel
