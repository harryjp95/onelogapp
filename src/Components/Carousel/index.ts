import Carousel from './Carousel'
export * from './'
export default Carousel