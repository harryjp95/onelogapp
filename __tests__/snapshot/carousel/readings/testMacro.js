import 'react-native'
import React from 'react'
import renderer from 'react-test-renderer'

import { MacroReading } from '../../../../src/Components/Carousel/Readings'

const testData = {
  kcal: 100,
  carbs: 10.0,
  sugar: 10.0,
  protein: 10.0,
  fat: 10.0
}

it('renders correctly', () => {
  const macroReading = renderer.create(<MacroReading data={testData} />)

  expect(macroReading).toMatchSnapshot()
})
