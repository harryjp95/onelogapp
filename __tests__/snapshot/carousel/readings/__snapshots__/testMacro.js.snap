// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ebebeb",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "space-around",
      "paddingBottom": 10,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "column",
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#3f3d3d",
          "fontSize": 16,
        }
      }
    >
      Kcal
    </Text>
    <Text
      style={
        Object {
          "color": "#3f3d3d",
          "fontSize": 16,
        }
      }
    >
      Carbs
    </Text>
    <Text
      style={
        Object {
          "color": "#3f3d3d",
          "fontSize": 16,
        }
      }
    >
      Sugar
    </Text>
    <Text
      style={
        Object {
          "color": "#3f3d3d",
          "fontSize": 16,
        }
      }
    >
      Protein
    </Text>
    <Text
      style={
        Object {
          "color": "#3f3d3d",
          "fontSize": 16,
        }
      }
    >
      Fat
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "column",
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "black",
          "fontSize": 16,
        }
      }
    >
      100.0
    </Text>
    <Text
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "black",
          "fontSize": 16,
        }
      }
    >
      10.0
    </Text>
    <Text
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "black",
          "fontSize": 16,
        }
      }
    >
      10.0
    </Text>
    <Text
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "black",
          "fontSize": 16,
        }
      }
    >
      10.0
    </Text>
    <Text
      style={
        Object {
          "alignSelf": "flex-end",
          "color": "black",
          "fontSize": 16,
        }
      }
    >
      10.0
    </Text>
  </View>
</View>
`;
